src/mono/mono/metadata/object-internals.h

```diff
@@ -1647,6 +1647,30 @@ typedef struct {
 	MonoProperty *prop;
 } CattrNamedArg;
 
+#ifdef ENABLE_NETCORE
+// Keep in sync with System.Runtime.Loader.AssemblyLoadContext.InternalState
+typedef enum {
+	ALIVE = 0,
+	UNLOADING = 1
+} MonoManagedAssemblyLoadContextInternalState;
+
+// Keep in sync with System.Runtime.Loader.AssemblyLoadContext
+typedef struct {
+	MonoObject object;
+	MonoObject *unload_lock;
```

Comment:
Would reorder this and put the runtime visible fields at the top, so the other fields don't need to be duplicated here.

Reply:
I guess thats not possible because its a partial class. In that case, this is going to be pretty brittle since fields can be added to the non-mono specific part.

Reply:
This can be avoided by not using a GC handle, and instead passing an opaque handle (the alc pointer). In that case, probably another API is needed to obtain the managed alc object from the handle.

Reply:
Yeah, that's why I just added the note to keep them in sync. The alternative is I guess creating a Mono-specific field that just gets updated every time the main field is, but that seems like it would introduce so many other potential issues that it's not worth it.

Reply:
The problem with using an opaque handle is then you have to worry about unloading. You'd either have to be exposing public refcounting mechanisms or give embedders a handle with a contract to return it... The gchandle seems less troublesome.

Reply:
Maybe the ref count is the way to go though, since that's happening anyway internally due to unloading and we should just make it public? I feel like the gchandle is much easier for people to deal with though, and realistically we're going to need a `mono_alc_from_gchandle` regardless.

Reply:
We could also return a gchandle for some internal object we create ourselves and whose layout we control, which references the real alc object. Looks a bit weird, but it should work.
Or, we could keep a alc object->alc mapping in a hash table, but it would keep the alc object alive unless it uses a weak handle or something.